[JSPI] Automatically add promising JSPI wrappers to thread entry points.

brendandahl opened the pull request:
Use a WebAssembly.Promising wrapper on the thread entry points to allow
JSPI to be used from worker threads. The code that invokes the wrapper
also adds async/await to handle this async call to entry point.

Fixes https://github.com/emscripten-core/emscripten/issues/22354

sbc100 commented on src/runtime_pthread.js:
```diff
@@ -90,7 +90,7 @@ if (ENVIRONMENT_IS_PTHREAD) {
   self.onunhandledrejection = (e) => { throw e.reason || e; };
 
-  function handleMessage(e) {
+  {{{ asyncIf(ASYNCIFY == 2) }}} function handleMessage(e) {
```
It seems like these fixes would apply to JSPI using in the main function of any pthread, not just the PROXY_TO_PTHREAD mode.   Is there something specific to PROXY_TO_PTHREAD here?

brendandahl replied:
Yeah, it's not totally specific to PROXY_TO_PTHREAD, more it showed up as an issue there since the entry point for the main proxy is not the users actual main. 

With a user defined entry point there were ways of handling this (see core.test_pthread_join_and_asyncify). I think I can actually change that test now to not require the user to specify that the entry function is JSPI. Hower, this makes me wonder if we should automatically JSPI wrap thread entry functions or require the user to specify that they need JSPI. FWIW, we currently auto JSPI wrap main.

sbc100 replied:
Do an alternative solution to this one would be to auto-wrap ` _main_thread` which is used in `_emscripten_proxy_main` ?

If the clang annotation was working we could use that there?

brendandahl replied:
> So an alternative solution to this one would be to auto-wrap ` _main_thread` which is used in `_emscripten_proxy_main` ?

Yeah, we'd need `EMSCRIPTEN_KEEPALIVE` or add it to `EXPORTED_FUNCTIONS` and also add it to `DEFAULT_ASYNCIFY_EXPORTS`.

> If the clang annotation was working we could use that there?

Yeah, I still need to fix binaryen to handle the annotation.

brendandahl commented:
Per discussion in meeting, we'll auto wrap all thread entry points. I've removed my new test, but updated the original test to cover the case of suspending on them main thread too. 

(sorry for the force push, but I wanted to amend to make sure I removed all new code)

kripken commented on test/core/test_pthread_join_and_asyncify.c:
```diff
@@ -13,15 +13,17 @@ EM_ASYNC_JS(int, async_call, (), {
 
 int main() {
   pthread_t id;
+  // Test that JSPI works on main thread.
```
```suggestion
  // Test that JSPI works on the main thread.
```

brendandahl commented:
Wasm64 was broken since it was trying to do `WebAssembly.promising(func)` where func was a wrapper and not the actual wasm export. I added an option to makeDynCall to add the `WebAssembly.promising` there instead.
